File: generate_voices.py
#!/usr/bin/env python3
"""
Generate robotic AI voices for workout activities and sentences using Kokoro TTS and Pedalboard effects.
"""

import os
import json
import subprocess
from pathlib import Path
from kokoro_onnx import Kokoro
# Added Bitcrush, Phaser, Compressor, HighpassFilter
from pedalboard import Pedalboard, Chorus, Reverb, PitchShift, HighpassFilter, Bitcrush, Phaser, Compressor, Gain
from pedalboard.io import AudioFile
import numpy as np


def apply_robotic_effect(audio_data: np.ndarray, sample_rate: int) -> np.ndarray:
    """
    Apply "Sci-Fi AI" voice effects.
    """
    board = Pedalboard([
        Phaser(rate_hz=2.0, depth=0.4, feedback=0.3, mix=0.4),
        Chorus(rate_hz=5.0, depth=0.1, centre_delay_ms=5.0, mix=0.3),
        Reverb(room_size=0.5, damping=0.5, wet_level=0.3, dry_level=0.7),
    ])
    
    # Apply effects
    processed_audio = board(audio_data, sample_rate)
    
    return processed_audio


def extract_activity_names(workout_dir: Path) -> set:
    """Extract unique activity names from all workout JSON files."""
    activity_names = set()
    
    for json_file in workout_dir.glob("*.json"):
        with open(json_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
            for activity in data.get('activities', []):
                name = activity.get('name')
                if name:
                    activity_names.add(name)
    
    return activity_names


def extract_sentences(sentences_file: Path) -> list:
    """Extract sentences from the sentences.json file."""
    if not sentences_file.exists():
        return []
    
    with open(sentences_file, 'r', encoding='utf-8') as f:
        data = json.load(f)
        return data if isinstance(data, list) else []


def sanitize_filename(name: str) -> str:
    """Convert activity name to safe filename."""
    return name.lower().replace(' ', '-').replace('_', '-')


def convert_wav_to_m4a(wav_path: Path, m4a_path: Path) -> bool:
    """
    Convert WAV to web-performant M4A using FFmpeg.
    Uses AAC codec with optimized settings for web playback.
    Removes the WAV file after successful conversion.
    
    Returns True if conversion was successful, False otherwise.
    """
    try:
        # FFmpeg command for web-optimized M4A
        # -c:a aac: Use AAC codec
        # -b:a 128k: Set bitrate to 128kbps (good balance of quality and size)
        # -movflags +faststart: Optimize for web streaming (moov atom at start)
        # -y: Overwrite output file if it exists
        cmd = [
            'ffmpeg',
            '-i', str(wav_path),
            '-c:a', 'aac',
            '-b:a', '128k',
            '-movflags', '+faststart',
            '-y',
            str(m4a_path)
        ]
        
        result = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            check=True
        )
        
        # Remove the WAV file after successful conversion
        wav_path.unlink()
        
        return True
    except subprocess.CalledProcessError as e:
        print(f"    X FFmpeg conversion failed: {e.stderr.decode('utf-8', errors='ignore')}")
        return False
    except Exception as e:
        print(f"    X Conversion error: {e}")
        return False


def main():
    """Main function to generate voices."""

    voice_name = "heart"
    
    # Define paths
    project_root = Path(__file__).parent
    workout_dir = project_root / "src" / "lib" / "data" / "workouts"
    sentences_file = project_root / "src" / "lib" / "data" / "sentences.json"
    activity_output_dir = project_root / "static" / "audio" / "voice" / voice_name
    sentence_output_dir = project_root / "static" / "audio" / "voice" / voice_name
    
    activity_output_dir.mkdir(parents=True, exist_ok=True)
    sentence_output_dir.mkdir(parents=True, exist_ok=True)
    
    print("Extracting activity names and sentences...")
    activity_names = extract_activity_names(workout_dir)
    sentences = extract_sentences(sentences_file)
    
    # Model files
    model_file = project_root / "kokoro-v1.0.onnx"
    voices_file = project_root / "voices-v1.0.bin"
    
    if not model_file.exists() or not voices_file.exists():
        print("ERROR: Model files not found in project root.")
        return
    
    print("\nInitializing Kokoro TTS engine...")
    kokoro = Kokoro(str(model_file), str(voices_file))
    voice = f"af_{voice_name}" 
    
    items_to_process = []
    
    for activity_name in sorted(activity_names):
        filename = sanitize_filename(activity_name)
        output_path = activity_output_dir / f"{filename}.m4a"
        items_to_process.append((activity_name, output_path, "activity", ""))
    
    for sentence in sentences:
        filename = sanitize_filename(sentence)
        output_path = sentence_output_dir / f"{filename}.m4a"
        items_to_process.append((sentence, output_path, "sentence", ""))
    
    # Add numbers 0-50 for counting
    for num in range(51):
        num_text = str(num)
        output_path = sentence_output_dir / f"{num}.m4a"
        items_to_process.append((num_text, output_path, "number", ""))
    
    print(f"\nProcessing {len(items_to_process)} items...\n")
    
    for text, output_path, item_type, prefix in items_to_process:
        
        # Skip if already exists
        if output_path.exists():
            print(f"  - Exists: {output_path.name}")
            continue
        
        try:
            print(f"  > Generating: {text}")

            # --- FIX FOR KOKORO INTONATION ---
            # Append a period to force 'statement' intonation.
            # Otherwise "Pushups" sounds like "Pushups?"
            text_to_speak = prefix + text.strip()
            if not text_to_speak.endswith(('.', '!', '?')):
                text_to_speak += "."

            audio_data, sample_rate = kokoro.create(
                text_to_speak,
                voice=voice,
                speed=1.0,
                lang='en-us'
            )
            
            if not isinstance(audio_data, np.ndarray):
                audio_data = np.array(audio_data, dtype=np.float32)
            
            if audio_data.ndim == 1:
                audio_data = audio_data.reshape(1, -1)
            elif audio_data.ndim == 2 and audio_data.shape[0] > audio_data.shape[1]:
                audio_data = audio_data.T

            silence_duration_sec = 0.5
            silent_samples_count = int(sample_rate * silence_duration_sec)

            num_channels = audio_data.shape[0]
            silence_array = np.zeros((num_channels, silent_samples_count), dtype=audio_data.dtype)

            audio_data_with_silence = np.concatenate([audio_data, silence_array], axis=1)
            
            # Apply fixed effects
            processed_audio = apply_robotic_effect(audio_data_with_silence, sample_rate)
            
            # Write to WAV first
            wav_path = output_path.with_suffix('.wav')
            with AudioFile(str(wav_path), 'w', sample_rate, num_channels=processed_audio.shape[0]) as f:
                f.write(processed_audio)
            
            # Convert WAV to web-performant M4A and remove WAV
            if not convert_wav_to_m4a(wav_path, output_path):
                print(f"    X Failed to convert {wav_path.name}")
                # Keep the WAV file if conversion failed
                continue
            
        except Exception as e:
            print(f"  X Error: {e}")
            continue
    
    print("\nDone.")

if __name__ == "__main__":
    main()

File: generate_voices.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { spawn } from 'child_process';
import { KokoroTTS } from 'kokoro-js';
import WavEncoder from 'wav-encoder';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuration
const VOICE_NAME = 'heart'; // Using 'heart' voice as in the Python script
const MODEL_ID = 'onnx-community/Kokoro-82M-ONNX';
const OUTPUT_DIR = path.join(__dirname, 'static', 'audio', 'voice', VOICE_NAME);

// Ensure output directory exists
if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

// Paths
const WORKOUT_DIR = path.join(__dirname, 'src', 'lib', 'data', 'workouts');
const SENTENCES_FILE = path.join(__dirname, 'src', 'lib', 'data', 'sentences.json');

/**
 * Sanitize text for use as a filename.
 */
function sanitizeFilename(text) {
    return text.toLowerCase().replace(/\s+/g, '-').replace(/_/g, '-').replace(/[^a-z0-9-&]/g, '');
}

/**
 * Extract unique activity names from workout JSON files.
 */
function extractActivityNames() {
    const activityNames = new Set();
    if (fs.existsSync(WORKOUT_DIR)) {
        const files = fs.readdirSync(WORKOUT_DIR).filter(file => file.endsWith('.json'));
        for (const file of files) {
            const filePath = path.join(WORKOUT_DIR, file);
            try {
                const data = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
                if (data.activities) {
                    for (const activity of data.activities) {
                        if (activity.name) {
                            activityNames.add(activity.name);
                        }
                    }
                }
            } catch (err) {
                console.error(`Error reading ${file}:`, err);
            }
        }
    }
    return Array.from(activityNames);
}

/**
 * Extract sentences from sentences.json.
 */
function extractSentences() {
    if (fs.existsSync(SENTENCES_FILE)) {
        try {
            const data = JSON.parse(fs.readFileSync(SENTENCES_FILE, 'utf-8'));
            return Array.isArray(data) ? data : [];
        } catch (err) {
            console.error('Error reading sentences.json:', err);
            return [];
        }
    }
    return [];
}

/**
 * Convert raw audio data to WAV buffer.
 */
async function encodeWav(audioData, sampleRate) {
    return WavEncoder.encode({
        sampleRate: sampleRate,
        channelData: [audioData] // Mono
    });
}

/**
 * Apply robotic effects and convert to M4A using FFmpeg.
 */
function processAudio(inputPath, outputPath) {
    return new Promise((resolve, reject) => {
        // FFmpeg filter chain to approximate the Python script's Pedalboard effects:
        // Phaser: rate=2.0, depth=0.4, feedback=0.3 (approx decay)
        // Chorus: rate=5.0, depth=0.1, delay=5ms, mix=0.3
        // Reverb: room=0.5, damp=0.5, wet=0.3

        // Note: FFmpeg filters are different from Pedalboard, so this is an approximation.
        // aphaser=in_gain=0.6:out_gain=0.6:delay=3.0:decay=0.3:speed=2.0:type=t
        // chorus=0.7:0.3:5:0.1:5:2 (in_gain, out_gain, delays, decays, speeds, depths)
        // aecho=0.8:0.88:60:0.4 (simple reverb/echo)

        const filterComplex = [
            'aphaser=in_gain=0.8:out_gain=0.6:delay=3.0:decay=0.3:speed=2.0:type=t',
            'chorus=0.7:0.3:5:0.1:5:2',
            'aecho=0.8:0.9:50:0.3',
            'volume=12.0'
        ].join(',');

        const args = [
            '-y', // Overwrite
            '-i', inputPath,
            '-filter_complex', filterComplex,
            '-c:a', 'aac',
            '-b:a', '128k',
            '-movflags', '+faststart',
            outputPath
        ];

        const ffmpeg = spawn('ffmpeg', args);

        ffmpeg.on('close', (code) => {
            if (code === 0) {
                resolve();
            } else {
                reject(new Error(`FFmpeg exited with code ${code}`));
            }
        });

        ffmpeg.on('error', (err) => {
            reject(err);
        });
    });
}

async function main() {
    console.log('Starting voice generation...');

    // 1. Initialize TTS
    console.log(`Loading model ${MODEL_ID}...`);
    const tts = await KokoroTTS.from_pretrained(MODEL_ID, {
        dtype: "fp32", // Use fp32 for compatibility
    });

    // 2. Gather items to process
    const items = [];

    // Activities
    const activities = extractActivityNames();
    console.log(`Found ${activities.length} activities.`);
    activities.forEach(name => {
        items.push({
            text: name,
            filename: sanitizeFilename(name),
            type: 'activity'
        });
    });

    // Sentences
    const sentences = extractSentences();
    console.log(`Found ${sentences.length} sentences.`);
    sentences.forEach(text => {
        items.push({
            text: text,
            filename: sanitizeFilename(text),
            type: 'sentence'
        });
    });

    // Numbers 0-50
    console.log('Adding numbers 0-50...');
    for (let i = 0; i <= 50; i++) {
        items.push({
            text: i.toString(),
            filename: i.toString(),
            type: 'number'
        });
    }

    console.log(`Total items to process: ${items.length}`);

    // 3. Process items
    for (const item of items) {
        const outputPath = path.join(OUTPUT_DIR, `${item.filename}.m4a`);

        if (fs.existsSync(outputPath)) {
            console.log(`Skipping (exists): ${item.filename}`);
            continue;
        }

        console.log(`Generating: ${item.text}`);

        try {
            // Add punctuation to force statement intonation if needed
            let textToSpeak = item.text.trim();
            if (!/[.!?]$/.test(textToSpeak)) {
                textToSpeak += '.';
            }

            // Generate audio
            // Kokoro-js returns raw audio data
            const audio = await tts.generate(textToSpeak, {
                voice: `af_${VOICE_NAME}`,
            });

            // audio is likely an object with { audio: Float32Array, sampling_rate: number }
            // or just the array depending on version. 
            // Based on typical transformers.js / kokoro-js patterns:
            const rawAudio = audio.audio;
            const sampleRate = audio.sampling_rate;

            // Save to temp WAV
            const wavBuffer = await encodeWav(rawAudio, sampleRate);
            const tempWavPath = path.join(OUTPUT_DIR, `${item.filename}_temp.wav`);
            fs.writeFileSync(tempWavPath, Buffer.from(wavBuffer));

            // Process with FFmpeg
            await processAudio(tempWavPath, outputPath);

            // Cleanup temp file
            fs.unlinkSync(tempWavPath);

        } catch (err) {
            console.error(`Error processing ${item.text}:`, err);
        }
    }

    console.log('Done!');
}

main().catch(console.error);
